src/data/transforms.py

```diff
@@ -1,169 +1,233 @@
     Args:
-        transforms (Box): The preprocessing and augmentation techniques applied to the data.
-
-    Returns:
-        transforms (Compose): The list of BaseTransform.
+         transforms (sequence of BaseTransform, optional): The preprocessing and augmentation techniques
+            applied to the data (default: None, do not apply any transform).
     """
-    if transforms is None:
-        return None
-
-    _transforms = []
-    for transform in transforms:
-        cls = getattr(src.data.transforms, transform.name)
-        kwargs = transform.get('kwargs')
-        _transforms.append(cls(**kwargs) if kwargs else cls())
-    transforms = Compose(_transforms)
-    return transforms
-
 
-class BaseTransform:
-    """The base class for all transforms.
-    """
+    def __init__(self, transforms=None):
+        if transforms is None:
+            self.transforms = tuple()
+        else:
+            if not all(isinstance(transform, BaseTransform) for transform in transforms):
+                raise TypeError('All of the transforms should be BaseTransform.')
+            self.transforms = tuple(transforms)
 
-    def __init__(self):
-        pass
+    def __call__(self, *imgs, **transforms_kwargs):
+        """
+        Args:
+            imgs (tuple of numpy.ndarray): The images to be transformed.
+            transforms_kwargs (dict): The runtime kwargs for each transforms.
 
-    def __call__(self, *imgs, **kwargs):
-        raise NotImplementedError
+        Returns:
+            imgs (tuple): The transformed images.
+        """
+        for transform in self.transforms:
+            kwargs = transforms_kwargs.get(transform.__class__.__name__, {})
+            transformed = kwargs.pop('transformed', tuple(True for _ in range(len(imgs))))
+            if len(transformed) != len(imgs):
+                raise ValueError('The number of the transformed should be the same as the images.')
+            if not all(_transformed in [True, False] for _transformed in transformed):
+                raise ValueError('All of the transformed should be either True or False.')
+            if not any(transformed):
+                raise ValueError('The transformed should not be all False.')
+
+            if all(transformed):
+                imgs = transform(*imgs, **kwargs)
+            else:
+                transformed_imgs, untransformed_imgs = self._split_imgs(imgs, transformed)
+                transformed_imgs = transform(*transformed_imgs, **kwargs)
+                imgs = self._reassemble_imgs(transformed_imgs, untransformed_imgs, transformed)
+        return imgs
 
     def __repr__(self):
-        return self.__class__.__name__
+        format_string = self.__class__.__name__ + '('
+        for transform in self.transforms:
+            format_string += '\n'
+            format_string += f'    {transform}'
+        format_string += '\n)'
+        return format_string
 
+    @classmethod
+    def compose(cls, transforms=None):
```

Comment:
where would the `compose` be called

Reply:
It's a factory method of `Compose`, so it can be called to create a `Compose` instance with `transforms (BoxList)`. I think when defining `MyDataset` is where it can be called, for example,
```python
class MyDataset(BaseDataset):
    def __init__(self, transforms, **kwargs):
        super().__init__(**kwargs)
        self.transforms = Compose.compose(transforms)
```

Reply:
Record in #27 .